src/devices/sound/setapcm.h

```diff
@@ -0,0 +1,80 @@
+// license:BSD-3-Clause
+// copyright-holders:Tomasz Slanina
+#ifndef MAME_SOUND_SETAPCM_H
+#define MAME_SOUND_SETAPCM_H
+
+#pragma once
+
+#include "dirom.h"
+
+//**************************************************************************
+//  TYPE DEFINITIONS
+//**************************************************************************
+
+// ======================> setapcm_device
+
+// TODO: unknown address bus width
+class setapcm_device : public device_t,
+					public device_sound_interface,
+					public device_rom_interface<32>
+{
+public:
+	void snd_w(offs_t offset, u16 data, u16 mem_mask = ~0);
+	u16 snd_r(offs_t offset);
+	void sndctrl_w(offs_t offset, u16 data, u16 mem_mask = ~0);
+	u16 sndctrl_r();
+
+protected:
+	setapcm_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, u32 clock);
+
+	// device-level overrides
+	virtual void device_start() override;
+	virtual void device_clock_changed() override;
+
+	// sound stream update overrides
+	virtual void sound_stream_update(sound_stream &stream, std::vector<read_stream_view> const &inputs, std::vector<write_stream_view> &outputs) override;
+
+	// device_rom_interface implementation
+	virtual void rom_bank_updated() override;
+
+	int m_max_voices = 16; // max voices
```

Comment:
Either make this a `const int` set in the constructor and allocate the `voice_t` array dynamically with the correct number of elements, or make the `setapcm_device` a class template and make the number of voices a template parameter.